sherpa/astro/data.py

```diff
@@ -2716,12 +2721,31 @@ def get_background(self,
 
         See Also
         --------
-        delete_background, set_background
+        delete_background, get_backgrounds, set_background
 
         """
         bkg_id = self._fix_background_id(id)
         return self._backgrounds.get(bkg_id)
 
+    def get_backgrounds(self) -> list[(IdType, DataPHA)]:
+        """Return all the background components.
+
+        .. versionadded:: 4.17.0
+
+        Returns
+        -------
+        bkg : list of (int or str, sherpa.astro.data.DataPHA)
+           All background identifiers and datasets associated with
+           this PHA. The list will be empty if there is no associated
+           background.
+
+        See Also
+        --------
+        delete_background, get_background, set_background
+
+        """
+        return list(self._backgrounds.items())
```

Comment:
I like this better than the previous use. However, I still wonder if this level of indirection is needed . This is a one-liner. Why not just use `data._background.items()` [note below] directly inline? (While they might be some special cases, where it needs to be used as a list, usually it's just iterated over and then `.items()` is fine - that's interable.

Obviously, this won't hurt performance or anything. I just personally find code that uses too many levels of indirection hard to read. I know that "In computer science, every problem can be solved with one more level of indirection" and there are valid use cases, but it seems overly complicated.

note: If that's the recommended patter, it should probably be made part of the official API and be called `backgrounds` instead of `_backgrounds`.

To be clear, I'm not going to veto this PR over this -- it's a valid design decision -- I'm just pointing out what I personally would prefer.

Reply:
Thinking this further- do we need `self.background_ids`? Isn't that just `self._background.keys()`? Can be simplify a lot more by just making this dict non-private?

Reply:
Last thought before I move to the next commit: If the pattern is just `data._background.items()` then we also don't need any extra tests that show that we really get back what we expect by calling the "get me the list of backgrounds" function because `data._background.items()`  is just Python and not Sherpa functionality - we can rely on the fact that `keys()` or `items()` works on a dict.

Reply:
So, `background_ids` is interestingly implemented, but it was designed a long time ago when we didn't have the certainty we now have (e.g. that ordering of keys is deterministic now). I wouldn't want to remove `background_ids` since it is an existing API and I don't see any benefit to removing it, but it could be cleaned up (but I'd rather do that in a later PR unless I gain a huge amount of enthusiasm and time).

For background information: this API was also designed to be somewhat language agnostic (as it had to allow working with S-Lang as well as Python) in which case having an explicit API rather than just relying on the Python dict interface was almost a necessity. We are obviously long gone from those requirements, but the old API still exists and I am not convinced of the use of replacing the API (as I don't want to break existing code, including within Sherpa itself), and is it worth having two APIs (explicit `get`/`set`/`delete` calls versus "use dictionary methods)?

Potential advantages to the indirection that the current design has is that we hide the internal implementation and we can check that the keys to the dict are IdType and the values are DataPHA objects when inserting them. We could instead overload the `backgrounds` property to allow this (but it gets a bit messy).

Reply:
So, with the current setup we can *technically* add a background to a dataset and then "remove" it - by setting the `background_ids` to an array without this identifier - but still have the background associated with the dataset.

I guess this would let you quickly answer the question "how much of a difference does background component xxx make", but I'm not sure it makes that much sense. This does, however, suggest that my `get_backgrounds` helper could break existing code (since it doesn't care about the `background_ids` field). It could be made to do so. Need to think.